bpf: allow to enable host_routing and endpoint routes simultaneously

aspsk opened the pull request:
What the title says, see the commit message for details

Fixes: #14240

<!-- Description of change -->

```release-note
Enable endpoint routes + veth fast redirect support
```

julianwiedmann commented on bpf/bpf_lxc.c:
```diff
@@ -973,19 +979,19 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
 		if (ep) {
-#ifdef ENABLE_ROUTING
 			if (ep->flags & ENDPOINT_F_HOST) {
+				if (is_defined(ENABLE_ROUTING)) {
```
Something that triggers me whenever I see this code - is the check for `ENDPOINT_F_HOST` actually needed? After we jump to `to_host`, we end up checking `*dst_id == HOST_ID` instead. Are the two conditions completely interchangeable?

If so we could also check for `HOST_ID` here, avoid the endpoint-lookup, and make the overall logic a bit simpler.

aspsk replied:
Good point. The `dst_id` is taken from the `ipcache` map, so could it happen that the entry doesn't exist for an existing endpoint?

I wouldn't try to patch this in the context of this PR in any case (also given that we're close to the freeze)

pchaigno commented on pkg/datapath/linux/config/config.go:
```diff
@@ -863,7 +863,7 @@ func (h *HeaderfileWriter) writeTemplateConfig(fw *bufio.Writer, e datapath.Endp
 	}
 
-	if option.Config.ForceLocalPolicyEvalAtSource {
+	if option.Config.ForceLocalPolicyEvalAtSource || (option.Config.EnableEndpointRoutes && !option.Config.EnableHostLegacyRouting) {
```
IIUC, this is okay and won't end up causing https://github.com/cilium/cilium/issues/14657 because with BPF Host Routing, we'll skip the second policy enforcement. Is that correct?

I think I would prefer to have that bit of logic in the datapath (so using the two corresponding macros) with a big comment to explain why we're doing that and why it's fine.

aspsk replied:
> IIUC, this is okay and won't end up causing #14657 because with BPF Host Routing, we'll skip the second policy enforcement. Is that correct?

Yes, we jump right into the pod namespace.

> I think I would prefer to have that bit of logic in the datapath (so using the two corresponding macros) with a big comment to explain why we're doing that and why it's fine.

```
/*
 * When the host routing is enabled we need to check policies at source.
 * Otherwise, as the skb will be delivered directly to pod's namespace,
 * the ingress policy (the cil_to_container BPF program) will be bypassed.
 */
#if defined(ENABLE_ENDPOINT_ROUTES) && defined(ENABLE_HOST_ROUTING)
#  ifndef FORCE_LOCAL_POLICY_EVAL_AT_SOURCE
#  define FORCE_LOCAL_POLICY_EVAL_AT_SOURCE
#  endif
#endif
```
Does this work for you @pchaigno ?

pchaigno replied:
Yep. Thanks!

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -451,19 +452,19 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
 		if (ep) {
-#ifdef ENABLE_ROUTING
 			if (ep->flags & ENDPOINT_F_HOST) {
+				if (is_defined(ENABLE_ROUTING)) {
```
Why is this not an #ifdef anymore?

aspsk replied:
I've copy-pasted this from the original version of this patch, where Daniel used this to make code more readable

pchaigno replied:
cc @borkmann

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -972,19 +978,19 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
 #endif
+					return DROP_HOST_UNREACHABLE;
+				}
+				goto pass_to_stack;
```
I'm trying to understand this change in behavior, but I'm not sure I understand how it even used to work. AFAICS, when endpoint routes were enabled, we used to deliver packets going to the host via `ipv4_local_delivery` but that will fail as `ep->lxc_id` is 0 for the host endpoint :confused:

borkmann replied:
> when endpoint routes were enabled, we used to deliver packets going to the host via ipv4_local_delivery

Don't we pass everything up the stack with per endpoint routes? The ipv{4,6}_local_delivery() was used for !per-endpoint routes case to deliver into Pods.

pchaigno replied:
Well, I guess that's part of the weird thing here. If you look at the code as it is in master today, you can see that the call to `ipv4_local_delivery` (3 lines below) was outside the `ENABLE_ROUTING` guard. So we were (and still are) calling `ipv4_local_delivery` regardless of the endpoint routes setting it seems :confused:

aspsk replied:
In master this was actually protected by the `is_defined(ENABLE_ROUTING) || hairpin`, so in case of endpoint routes this piece of code is skipped

pchaigno replied:
Oh, right :facepalm: I was only looking at the ifdefs.

I think I get the change now. The old check on `ENABLE_ROUTING` (line 975) was useless then, right?

aspsk replied:
The old one was to optimize code for the hairpin case, I guess

pchaigno commented on bpf/lib/fib.h:
```diff
@@ -10,6 +10,12 @@
+ * Translation into human language: do a fib lookup if there is more than one
+ * output device or per-endpoint routes are enabled (which => !ENABLE_ROUTING)
+ */
+#define DO_FIB_LOOKUP (!is_defined(ENABLE_SKIP_FIB) || !is_defined(ENABLE_ROUTING))
```
I'd make it `ENABLE_FIB_LOOKUP` if possible, to be consistent with other macros.

pchaigno commented on bpf/lib/fib.h:
```diff
@@ -10,6 +10,12 @@
 
+/*
+ * Translation into human language: do a fib lookup if there is more than one
+ * output device or per-endpoint routes are enabled (which => !ENABLE_ROUTING)
```
Can we get the *why* instead of the *what*? :grin: 
I'm currently missing why your patch needs to change this.

aspsk replied:
@pchaigno I've replaced this with a proper macro

julianwiedmann replied:
+1 to what @pchaigno said - this needs an explanation *why* per-EP routes now always requires a FIB lookup.

pchaigno reviewed (commented):
Couple comments below. All in all, this is tricky logic so I think it warrants a few more explanations :slightly_smiling_face:

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -956,11 +957,16 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
+	 *
+	 * For the case where ENABLE_ROUTING is disabled, but the fast redirect
+	 * enabled we do lookup the local endpoint here to check whether we
+	 * must pass the packet up the stack for the host itself.
```
That's not the only reason to lookup the local endpoint, no? We also want to run through `ipv4_local_delivery` to enforce ingress policies for that endpoint.

aspsk replied:
Yes, thanks, I will update this comment

julianwiedmann commented on bpf/bpf_lxc.c:
```diff
@@ -973,19 +981,19 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
 #endif
+					return DROP_HOST_UNREACHABLE;
+				}
+				goto pass_to_stack;
```
For `ENDPOINT_F_HOST` and per-EP routes, this jump to `pass_to_stack` means we're bypassing quite a bit of code. Can you explain why it's ok to bypass the [section right below](https://github.com/cilium/cilium/blob/fcf2b7ad1c47cc7309e752eb087ecfbda7ff8a6e/bpf/bpf_lxc.c#L992-L1001) (Host-Ingress Policy Enforcement with HostFW + per-EP routes)? If it's no longer relevant, can it be removed?

aspsk replied:
This was addressed by moving host firewall check above

borkmann commented on bpf/lib/l3.h:
```diff
@@ -14,6 +14,17 @@
+#  ifndef FORCE_LOCAL_POLICY_EVAL_AT_SOURCE
+#  define FORCE_LOCAL_POLICY_EVAL_AT_SOURCE
+#  endif
+#endif
```
I would move this section out of the C code and into the agent. The agent could enable option.Config.ForceLocalPolicyEvalAtSource when it knows that option.Config.EnableEndpointRoutes is true and option.Config.EnableHostLegacyRouting is false. I'd also move the above comment to that code location where you set this in the agent. Less ifdef'ery then for the C code.

borkmann commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -449,8 +449,6 @@ func finishKubeProxyReplacementInit() error {
 		case option.Config.IptablesMasqueradingEnabled():
 			msg = fmt.Sprintf("BPF host routing requires %s.", option.EnableBPFMasquerade)
 		// All cases below still need to be implemented ...
```
nit: I think this comment above could then also be removed. 🎉

borkmann commented on bpf/bpf_lxc.c:
```diff
@@ -527,8 +527,20 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
+		tail_call_static(ctx, &POLICY_CALL_MAP, HOST_EP_ID);
+		return DROP_MISSED_TAIL_CALL;
+	}
+#endif /* ENABLE_HOST_FIREWALL && !ENABLE_ROUTING */
```
nit: Could we split out this move for HF above from the main commit for v4/v6 bits? Would be nice to move this out so the main commit shrinks a bit more, and for the HF we just do the move as a prep commit with no change in functionality.

borkmann reviewed (approved):
LGTM, just few nits below that would be nice to address before merging.

aspsk committed:
bpf: bpf_lxc: swap some code blocks (non-functional)

Move processing of host firewall one block above. This doesn't change
functionality, as the following block is executed only if ENABLE_ROUTING or
hairpin_flow is set and the host policy will be checked only if !ENABLE_ROUTING
and the destination is host (which is not the case for hairpin_flow).

This commit is required to simplify later changes.

Signed-off-by: Anton Protopopov <aspsk@isovalent.com>

aspsk committed:
bpf: allow to enable host routing and endpoint routes simultaneously

We have a long-term goal to enable endpoint routes by default. This patch
removes one of the blockers, namely, that we can not enable endpoint routes
together with optimized routing (bpf host routing).

In order to enforce ingress policy for endpoints, when the endpoint routes are
enabled, we set ForceLocalPolicyEvalAtSource=true. At the moment this is the
only way to enforce the ingress policy when the bpf_redirect_peer() function is
used, because it bypasses the host-side of pod veth pair. [To use the
bpf_redirect() function here doesn't make sense, as this basically reverts all
the optimizations made by the bpf_redirect_peer() function]. ["In future" we
will be able to force the ingress policy at the destination, because the new
"meta" networking device will have a program attached in the pod namespace, and
the bpf_redirect_peer() helper won't bypass it.]

We still require the cil_to_container program attached to the lxc host part
egress, because we have traffic which is generated by host, so that we can
apply policies to it.

This should work, and policies should be applied once, because the
ipv{4,6}_policy function will eventually execute the redirect_ep() function for
endpoints, which will trigger the bpf_redirect_peer() function, which will
bypass the cil_to_container program, and thus policy will be applied once. This
way policies are applied + we optimize redirects.

In case of endpoint routes & host routing the IPv6 breaks, so enable host
routing only for IPv4 case.  One reason is that the host can't populate
neighbour entries for endpoints.  This happens because in case of endpoint
routes ICMPv6 messages originate on [host] lxc veth devices. Those devices are
only configured with IPv6 link-local addresses, so the ICMPv6 answer from the
endpoint is sent to a link-local address which is recognized "world" by cilium,
but fib_redirect can't redirect it (as it should be delivered to stack, not
forwarded). In IPv4 case this is not a problem becase the lxc devices aren't
configured with L3 addresses and thus a legitimate host IP address is chosen.

This commit is based on the original PR from Daniel Borkmann:

    https://github.com/cilium/cilium/pull/14382

Fixes: #14240

Signed-off-by: Anton Protopopov <aspsk@isovalent.com>

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -563,7 +563,8 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
 #endif /* ENABLE_HOST_FIREWALL && !ENABLE_ROUTING */
```
From commit description:
> In order to enforce Host Firewall (CCNP), we move a corresponding code section
in bpf_lxc.c one block above.

I believe this needs to be updated because it was moved to its own commit (:pray:)

pchaigno commented on pkg/datapath/linux/config/config.go:
```diff
@@ -874,7 +874,9 @@ func (h *HeaderfileWriter) writeTemplateConfig(fw *bufio.Writer, e datapath.Endp
+	// When both the host routing and endpoint routes are enabled we need
+	// to check policies at source.
+	if option.Config.ForceLocalPolicyEvalAtSource || (option.Config.EnableEndpointRoutes && !option.Config.EnableHostLegacyRouting) {
 		fmt.Fprintf(fw, "#define FORCE_LOCAL_POLICY_EVAL_AT_SOURCE 1\n")
```
Setting this in the global datapath configuration will force policy evaluation at the source for lxc -> lxc, but also for hostns -> lxc, eth0 -> lxc, and overlay -> lxc because all of those paths call `ipv{4,6}_local_delivery`. Is that really what we want?

AFAICS, that will cause us again to enforce policies twice for all of those paths, when endpoint routes and BPF Host Routing are enabled.

aspsk replied:
Thanks for the comments @pchaigno!

The idea is that `!option.Config.EnableHostLegacyRouting` means that after checking policy at source we will forward the packet directly to the pod using `redirect_peer`, thus bypassing the outer part of the veth. The only case when we trigger the policy program at the outer veth is when traffic gets there directly from the hostns (as endpoint routes are set). Does this make sense for you?

pchaigno replied:
Ok, I think I see what I was missing. The packet will then follow the path: bpf_lxc:handle_policy() -> tail_ipv4_ct_ingress_policy_only() -> tail_ipv4_policy() -> ipv4_policy() -> common.h:redirect_ep() -> ctx_redirect_peer(). And for hostns -> lxc, it won't skip the destination's ingress BPF program because of the `needs_backlog` variable, is that right?

Overall, it feels super fragile. The path is complex, with many conditions that could send us into a double enforcement or policy bypass if they ever change the wrong way.

What kind of test coverage do we have for this? Will we notice if `FORCE_LOCAL_POLICY_EVAL_AT_SOURCE` disappears tomorrow (consequence: policy bypass), or if `from_host` is not correctly propagated into `needs_backlog` anymore (consequence: policy bypass), or if [this condition](https://github.com/cilium/cilium/blob/4310b5e5c4a15049d963cbad8791226567743e9c/bpf/bpf_lxc.c#L1881) is extended (consequence: double enforcement)?

aspsk replied:
Hi @pchaigno! We will notice if policy is not applied correctly via the connectivity test. However, we currently don't automatically test dual policy enforcement.

pchaigno replied:
That will only be the case if this configuration is covered in some test somewhere. Is it? Do the Datapath Conformance tests enable BPF Host Routing?

aspsk replied:
The host routing is enabled by default unless it is disabled on kernels which don't support it (= `EnableHostLegacyRouting` is disabled by default.) The net-next, at least, runs with the host routing enabled.

pchaigno replied:
Discussed offline. Today the net-next tests automatically disable BPF Host Routing because it is not supported. Once this PR is merged, this "auto-disablement" won't happen anymore and the tests will start testing BPF Host Routing with endpoint routes without any changes to the tests themselves being needed.

LGTM.

pchaigno reviewed (changes requested):
I might be missing something, but AFAICS, this will reintroduce bug https://github.com/cilium/cilium/issues/14657. Please see comment below.

There's also one comment from Julian that wasn't resolved/answered above.

zacharysarah commented on bpf/bpf_lxc.c:
```diff
@@ -1003,12 +1008,30 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
-	 * been passed to the stack.
+	 * disabled (e.g. the case in per-endpoint routes). Otherwise, the
+	 * packet will be dropped by the kernel if it is going to be routed via
+	 * an interface it came from after it has been passed to the stack.
```
Avoid Latin abbreviations.
Tighten up verb tenses for clarity. (Please review for technical accuracy.)

```suggestion
	 * disabled (for example, with per-endpoint routes). Otherwise, the
	 * packet will be dropped by the kernel if the packet will be routed through
	 * the interface it came from after the packet has been passed to the stack.
```

zacharysarah commented on bpf/bpf_lxc.c:
```diff
@@ -1003,12 +1008,30 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
+	 * enabled we do lookup the local endpoint here to check whether we
+	 * must pass the packet up the stack for the host itself. We also want
+	 * to run through ipv4_local_delivery to enforce ingress policies for
+	 * that endpoint.
```
Who is "we" here in both instances? It's unclear which component(s) are acting.

```suggestion
	 * If ENABLE_ROUTING is disabled but fast redirect is
	 * enabled, we do lookup the local endpoint here to check whether we
	 * must pass the packet up the stack for the host itself. We also want
	 * to run through `ipv4_local_delivery` to enforce ingress policies for
	 * that endpoint.
```

zacharysarah commented on bpf/lib/l3.h:
```diff
@@ -14,6 +14,17 @@
+/*
+ * When the host routing is enabled we need to check policies at source.
+ * Otherwise, as the skb will be delivered directly to pod's namespace,
+ * the ingress policy (the cil_to_container BPF program) will be bypassed.
```
Who is "we" here? Which component(s) are checking policies at source?
Avoid future tense.
```suggestion
 * When the host routing is enabled we need to check policies at source.
 * Otherwise, the skb is delivered directly to pod's namespace and
 * the ingress policy (the cil_to_container BPF program) is bypassed.
```

zacharysarah commented on pkg/datapath/linux/config/config.go:
```diff
@@ -874,7 +874,9 @@ func (h *HeaderfileWriter) writeTemplateConfig(fw *bufio.Writer, e datapath.Endp
 
-	if option.Config.ForceLocalPolicyEvalAtSource {
+	// When both the host routing and endpoint routes are enabled we need
+	// to check policies at source.
```
Who is "we"?

zacharysarah reviewed (changes requested):
@aspsk Good start! I have some questions about which component(s) are acting in specific instances.

pchaigno reviewed (changes requested):
Blocking on above testing discussion. This is fairly fragile, so we really need to make sure this is covered somewhere in CI. AFAICS, that is not the case today.

It may be as simple as extending the Conformance Datapath workflow.

pchaigno reviewed (approved):
Above discussion is resolved. LGTM!
